### PR TITLE
[Relax][Op] register gradients for some operators

### PR DESCRIPTION
**Pull request by SiriusNEO**
This PR registers gradients using `relay.op.register_gradient` for the following relax operators:
- `add`
- `subtract`
- `transpose`
- `relu`
- `nn.matmul`
- `nn.softmax`
- `nn.cross_entropy`
- `nn.softmax_cross_entropy`
- `sigmoid`
- `tanh`

Numeric tests are passed for these implementations. (Running this test may take a period of time, approximately 1~2 min).

### CODE REVIEW

File: python/tvm/relax/op/gradient.py [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+	sum,
+	less,
+	where,
+	collapse_sum_to,
+	log,
+	multiply,
+	divide,
+	negative,
+	subtract,
+	transpose,
```
Comment (MasterJH5574) [inline python/tvm/relax/op/gradient.py]: Oh we should always use whitespaces for indent, instead of tab.

File: python/tvm/relax/op/gradient.py [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+	sigmoid,
+	tanh,
+	ones,
+	zeros
+)
+
+
+@register_gradient("relax.add")
+def add_grad(orig, grad):
```
Comment (MasterJH5574) [inline python/tvm/relax/op/gradient.py]: Could we add proper type annotation to the parameters? Like
```suggestion
def add_grad(orig: AAA, grad: BBB):
```

File: python/tvm/relax/op/gradient.py [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+"""Gradient definitions for Relax operators"""
+from __future__ import annotations
```
Comment (MasterJH5574) [inline python/tvm/relax/op/gradient.py]: Does this file rely on this import? It seems to me that this import is redundant.

Comment (Hzfengsy) [inline python/tvm/relax/op/gradient.py]: This line is no longer needed IIRC

File: tests/python/relax/test_numeric_gradient.py [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+from tvm.relax.transform.op_legalizer import OperatorLegalizer
+from tvm.testing.utils import check_numerical_grads
+from tvm.ir.op import Op
+import tvm.relax.op.gradient
+
+
+def relax_check_gradients(op_func, op_name, input_numpy, target, dev, output_shape):
+    """
+        Generate module and run it to check numberic gradients.
+    """
```
Comment (Hzfengsy) [inline tests/python/relax/test_numeric_gradient.py]:
```suggestion
    """Generate module and run it to check numberic gradients."""
```

File: python/tvm/relax/op/gradient.py [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+def multiply_grad(orig, grad):
+	"""Returns [grad * y, grad * x]."""
+	x, y = orig.args
+	return [collapse_sum_to(multiply(grad, y), x.shape), collapse_sum_to(multiply(grad, x), y.shape)]
+
+
+@register_gradient("relax.transpose")
+def transpose_grad(orig, grad):
+	"""Returns grad transposed over the complement of original transpose axes"""
+	"""TODO: Do not support more than one dimensions"""
```
Comment (MasterJH5574) [inline python/tvm/relax/op/gradient.py]: I’m curious about this “to do”. Do you mean this gradient operator only supports the case of transposing a 2-dim tensor? This transpose operator will reverse the axis order by default if we don’t specify the `axes` field.

File: tests/python/relax/test_numeric_gradient.py [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+def test_relu(target, dev):
+    data1_numpy = np.random.uniform(-1, 1, (16, 16)).astype(np.float32)
+    relax_check_gradients(relax.op.nn.relu, "relax.nn.relu", [data1_numpy], target, dev, (16, 16))
+
+
+@tvm.testing.parametrize_targets("llvm")
+def test_matmul(target, dev):
+    data1_numpy = np.random.randint(0, 16, (7, 8)).astype(np.float32)
+    data2_numpy = np.random.randint(0, 16, (8, 10)).astype(np.float32)
+    relax_check_gradients(relax.op.nn.matmul, "relax.nn.matmul", [data1_numpy, data2_numpy], target, dev, (7, 10))
```
Comment (MasterJH5574) [inline tests/python/relax/test_numeric_gradient.py]: The matmul operator supports high-dimensional input tensors, like `(5, 4, 1, 8, 9) x (4, 6, 9, 10) => (5, 4, 6, 8, 10)`. I think we would better test if the registered gradient works for high-dimensional input cases as well.

File: python/tvm/relax/op/gradient.py (hunk 2 above)
Comment (SiriusNEO) [reply python/tvm/relax/op/gradient.py]: Seems that `relay/op/_tensor_grad.py` also doesn't has type annotations?

Comment (MasterJH5574) [reply python/tvm/relax/op/gradient.py]: You can do better than them :-)

**Review by MasterJH5574 — APPROVED** [review]

File: python/tvm/relax/op/gradient.py [hunk 7]
```diff
@@ -0,0 +1,191 @@
...
+@register_gradient("relax.sum")
+def sum_grad(orig: Call, grad: Var):
+    """Returns [grad * ones(x.shape)]."""
+    return [multiply(grad, ones(orig.args[0].shape))]
+
+
+@register_gradient("relax.nn.softmax")
+def softmax_grad(orig: Call, grad: Var):
+    """Gradient of softmax."""
+    return [multiply(subtract(grad, sum(multiply(grad, orig), orig.attrs.axis, True)), orig)]
```
Comment (MasterJH5574) [inline python/tvm/relax/op/gradient.py]: Ah I missed these two operators. They all have the attribute `axis` which you may also need to consider.

**Comment by SiriusNEO** [comment]
@MasterJH5574 Hi Ruihang, `sum` and `softmax` with different attributes are resolved in the lastest commit. Besides, after rebasing I found one of my test failed. And I found this is because `op_legalizer` does not handle the negative axis case in `_mean`. I tried to fix it as well. Also, I found a typo in `transform.cc` and fixed it.

**Comment by Ubospica** [comment]
The bug we found can be reproduced on this script:
```
from tvm import relax

x = relax.Var("x", [2, 3, 4], relax.DynTensorType(ndim=3, dtype="float32"))
bb = relax.BlockBuilder()
with bb.function("main", [x]):
    gv = bb.emit(R.ones_like(x))
    bb.emit_func_output(gv)

bb.get()["main"].show()
```
The name of the printed function is "foo" instead of "main".

The problem we identify is that `RelaxScriptPrinter` would not update the name of function `relax_func_name_` when printing relax functions, but in printing IRModules. See https://github.com/mlc-ai/relax/blob/f749fb2d6a9cacc9264405effdc1bb5072deac99/src/printer/relax_script_printer.cc#L350-L360

And

https://github.com/mlc-ai/relax/blob/f749fb2d6a9cacc9264405effdc1bb5072deac99/src/printer/relax_script_printer.cc#L529-L544

If we pass function as input, `relax_func_name_ ` will keep its default value "foo".

This bugfix adds an assignment to relax_func_name_  for printing function.

**Comment by MasterJH5574** [comment]
@Ubospica For your reproducible code at the beginning, I suppose `bb.get()["main"]` does not have the `global_symbol`.

Since https://github.com/tlc-pack/relax/commit/fed6b8fc88b824ec68260417793447dbe524c4c3, BlockBuilder no longer actively adds the `global_symbol` attribute to either Relax Function and TIR PrimFunc. So in your case, `bb.get()["main"]` should be a Function without `global_symbol` attribute. And it makes sense to print the default function name when there is no additional information.

If a Function happens to have `global_symbol`, I think the TVMScript printer now will print out the attribute explicitly: https://github.com/mlc-ai/relax/blob/f749fb2d6a9cacc9264405effdc1bb5072deac99/src/printer/relax_script_printer.cc#L598-L608

This is the same for printing TIR PrimFuncs: when printing a single PrimFunc, the function name will be kept as the default one and the `global_symbol` attribute will be printed explicitly via `T.func_attr({...})`.

**Review by MasterJH5574 — APPROVED** [review]
